sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapi/AADOBOOAuth2AuthorizedClientProvider.java

```diff
@@ -0,0 +1,170 @@
+
+
+    private final Clock clock = Clock.systemUTC();
+
+    private final Duration clockSkew = Duration.ofSeconds(60);
+
+
+    @Override
+    public OAuth2AuthorizedClient authorize(OAuth2AuthorizationContext context) {
+        Assert.notNull(context, "context cannot be null");
+        ClientRegistration clientRegistration = context.getClientRegistration();
+        if (!AADAuthorizationGrantType.ON_BEHALF_OF.getValue().equals(clientRegistration.getAuthorizationGrantType().getValue())) {
+            return null;
+        }
+
+        OAuth2AuthorizedClient authorizedClient = context.getAuthorizedClient();
+        if (authorizedClient != null && !hasTokenExpired(authorizedClient.getAccessToken())) {
+            // If client is already authorized but access token is NOT expired than no
+            // need for re-authorization
+            return null;
+        }
+
+        return loadOboAuthorizedClient(context.getClientRegistration(), context.getPrincipal());
+    }
+
+    private boolean hasTokenExpired(AbstractOAuth2Token token) {
+        return this.clock.instant().isAfter(token.getExpiresAt().minus(this.clockSkew));
+    }
+
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private <T extends OAuth2AuthorizedClient> T loadOboAuthorizedClient(ClientRegistration clientRegistration,
+                                                                         Authentication principal) {
+
+        if (!(principal instanceof AbstractOAuth2TokenAuthenticationToken)) {
+            throw new IllegalStateException("Unsupported token implementation " + principal.getClass());
+        }
+        try {
+            String accessToken = ((AbstractOAuth2TokenAuthenticationToken<?>) principal).getToken()
+                .getTokenValue();
+            OnBehalfOfParameters parameters = OnBehalfOfParameters
+                .builder(clientRegistration.getScopes(), new UserAssertion(accessToken))
+                .build();
+            ConfidentialClientApplication clientApplication = createApp(clientRegistration);
+            if (null == clientApplication) {
+                return null;
+            }
+            String oboAccessToken = clientApplication.acquireToken(parameters).get().accessToken();
+            JWT parser = JWTParser.parse(oboAccessToken);
+            Date iat = (Date) parser.getJWTClaimsSet().getClaim("iat");
+            Date exp = (Date) parser.getJWTClaimsSet().getClaim("exp");
+            OAuth2AccessToken oAuth2AccessToken = new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER,
+                oboAccessToken,
+                Instant.ofEpochMilli(iat.getTime()),
+                Instant.ofEpochMilli(exp.getTime()));
+            OAuth2AuthorizedClient oAuth2AuthorizedClient = new OAuth2AuthorizedClient(clientRegistration,
+                principal.getName(),
+                oAuth2AccessToken);
+            LOGGER.info("load obo authorized client success");
+            return (T) oAuth2AuthorizedClient;
```

Comment:
Now the oAuth2AuthorizedClient will be stored to the repo/service, with key (clientRegistrationId, principal name). So will this be a problem for us?

Reply:
I will debug to test it today

Reply:
great

Reply:
@ZhuXiaoBing-cn could you please help confirm the name of the principal retrieved via the client credential flow? My concern is the client credential flow doesn't have a user, hence the sub should not represent a user.

Reply:
@saragluna the principal name used in the client credential flow is `anonymousUser`, which is provided by spring security by default.